In [1]:
import numpy as np
import my_submission as ms
import matplotlib.pyplot as plt

Using TensorFlow backend.


In [2]:
#from tensorflow.python.client import device_lib
#print(device_lib.list_local_devices())

In [3]:
#ms.save_mnist_dataset()
# Loads the already preprocessed dataset.
#(x_train, y_train), (x_test, y_test) = ms.load_mnist_dataset()
#print('x_train : ', x_train.shape, x_train.dtype)
#print('y_train : ', y_train.shape, y_train.dtype)
#print('x_test : ', x_test.shape, x_test.dtype)
#print('y_test : ', y_test.shape, y_test.dtype)

#x_train, y_train, x_test, y_test = ms.preprocess_mnist_dataset(x_train, y_train, x_test, y_test)
# Saves the preprocessed dataset locally as a .npz file.
#np.savez('mnist_dataset.npz', x_train = x_train, y_train = y_train, x_test = x_test, y_test = y_test)


In [4]:
# Verifying that the split has been done correctly
#print("Unique numbers in train set (shouldn't be any of the 0,1,8 or 9):")
#unique, counts = np.unique(y_train, return_counts=True)
#dict(zip(unique, counts))

In [5]:
# Verifying that the split has been done correctly
#print("Unique numbers in test set:")
#unique, counts = np.unique(y_test, return_counts=True)
#dict(zip(unique, counts))

In [6]:
#ms.build_CNN(x_train, y_train, x_test, y_test)
ms.siamese_network()

input_trainset:  (33149, 28, 28, 1)
target_trainset:  (33149,)
input_testset1:  (8288, 28, 28, 1)
target_testset1:  (8288,)
input_testset2:  (28563, 28, 28, 1)
target_testset2:  (28563,)
input_testset3:  (36851, 28, 28, 1)
target_testset3:  (36851,)
set1_digits [2, 3, 4, 5, 6, 7]
--------------------
training_pairs:  (61080, 2, 28, 28, 1)
training_target:  (61080,)
--------------------
test_pairs_set1:  (14652, 2, 28, 28, 1)
test_target_set1:  (14652,)
--------------------
test_pairs_set2:  (54592, 2, 28, 28, 1)
test_target_set2:  (54592,)
--------------------
test_pairs_set3:  (24420, 2, 28, 28, 1)
test_target_set3:  (24420,)
--------------------
Train on 61080 samples, validate on 14652 samples
Epoch 1/10
61080/61080 [==============================] - 335s 5ms/step - loss: 0.5539 - accuracy: 0.8559 - val_loss: 0.4450 - val_accuracy: 0.9160
Epoch 2/10
61080/61080 [==============================] - 292s 5ms/step - loss: 0.4479 - accuracy: 0.9147 - val_loss: 0.4529 - val_accuracy: 0.913

<h1>Results:</h1>
<p>
    input_trainset:  (33149, 28, 28, 1)
target_trainset:  (33149,)
input_testset1:  (8288, 28, 28, 1)
target_testset1:  (8288,)
input_testset2:  (28563, 28, 28, 1)
target_testset2:  (28563,)
input_testset3:  (36851, 28, 28, 1)
target_testset3:  (36851,)
set1_digits [2, 3, 4, 5, 6, 7]
--------------------
training_pairs:  (61080, 2, 28, 28, 1)
training_target:  (61080,)
--------------------
test_pairs_set1:  (14652, 2, 28, 28, 1)
test_target_set1:  (14652,)
--------------------
test_pairs_set2:  (54592, 2, 28, 28, 1)
test_target_set2:  (54592,)
--------------------
test_pairs_set3:  (24420, 2, 28, 28, 1)
test_target_set3:  (24420,)
--------------------
Train on 61080 samples, validate on 14652 samples
Epoch 1/10
61080/61080 [==============================] - 335s 5ms/step - loss: 0.5539 - accuracy: 0.8559 - val_loss: 0.4450 - val_accuracy: 0.9160
Epoch 2/10
61080/61080 [==============================] - 292s 5ms/step - loss: 0.4479 - accuracy: 0.9147 - val_loss: 0.4529 - val_accuracy: 0.9131
Epoch 3/10
61080/61080 [==============================] - 298s 5ms/step - loss: 0.4393 - accuracy: 0.9204 - val_loss: 0.4395 - val_accuracy: 0.9204
Epoch 4/10
61080/61080 [==============================] - 300s 5ms/step - loss: 0.4338 - accuracy: 0.9236 - val_loss: 0.4344 - val_accuracy: 0.9233
Epoch 5/10
61080/61080 [==============================] - 295s 5ms/step - loss: 0.4323 - accuracy: 0.9239 - val_loss: 0.4308 - val_accuracy: 0.9255
Epoch 6/10
61080/61080 [==============================] - 300s 5ms/step - loss: 0.4314 - accuracy: 0.9248 - val_loss: 0.4306 - val_accuracy: 0.9252
Epoch 7/10
61080/61080 [==============================] - 299s 5ms/step - loss: 0.4305 - accuracy: 0.9253 - val_loss: 0.4289 - val_accuracy: 0.9262
Epoch 8/10
61080/61080 [==============================] - 298s 5ms/step - loss: 0.4278 - accuracy: 0.9267 - val_loss: 0.4316 - val_accuracy: 0.9247
Epoch 9/10
61080/61080 [==============================] - 296s 5ms/step - loss: 0.4275 - accuracy: 0.9271 - val_loss: 0.4324 - val_accuracy: 0.9254
Epoch 10/10
61080/61080 [==============================] - 302s 5ms/step - loss: 0.4278 - accuracy: 0.9270 - val_loss: 0.4279 - val_accuracy: 0.9270
* Accuracy on training set: 92.95%
* Accuracy on test set 1: 92.70%
* Accuracy on test set 2: 61.19%
* Accuracy on test set 3: 77.70%
</p>



<p>
    input_trainset:  (33149, 28, 28, 1)
target_trainset:  (33149,)
input_testset1:  (8288, 28, 28, 1)
target_testset1:  (8288,)
input_testset2:  (28563, 28, 28, 1)
target_testset2:  (28563,)
input_testset3:  (36851, 28, 28, 1)
target_testset3:  (36851,)
set1_digits [2, 3, 4, 5, 6, 7]
--------------------
training_pairs:  (60936, 2, 28, 28, 1)
training_target:  (60936,)
--------------------
test_pairs_set1:  (14796, 2, 28, 28, 1)
test_target_set1:  (14796,)
--------------------
test_pairs_set2:  (54592, 2, 28, 28, 1)
test_target_set2:  (54592,)
--------------------
test_pairs_set3:  (24660, 2, 28, 28, 1)
test_target_set3:  (24660,)
--------------------
Train on 60936 samples, validate on 14796 samples
Epoch 1/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.4415 - accuracy: 0.8878 - val_loss: 0.3498 - val_accuracy: 0.9360
Epoch 2/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3350 - accuracy: 0.9433 - val_loss: 0.3211 - val_accuracy: 0.9514
Epoch 3/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3262 - accuracy: 0.9490 - val_loss: 0.3184 - val_accuracy: 0.9544
Epoch 4/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3187 - accuracy: 0.9525 - val_loss: 0.3278 - val_accuracy: 0.9484
Epoch 5/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3155 - accuracy: 0.9544 - val_loss: 0.3135 - val_accuracy: 0.9561
Epoch 6/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3115 - accuracy: 0.9564 - val_loss: 0.3099 - val_accuracy: 0.9576
Epoch 7/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3094 - accuracy: 0.9577 - val_loss: 0.3134 - val_accuracy: 0.9573
Epoch 8/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3083 - accuracy: 0.9584 - val_loss: 0.3114 - val_accuracy: 0.9569
Epoch 9/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3057 - accuracy: 0.9596 - val_loss: 0.3131 - val_accuracy: 0.9567
Epoch 10/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3058 - accuracy: 0.9598 - val_loss: 0.3177 - val_accuracy: 0.9545
Epoch 11/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3040 - accuracy: 0.9606 - val_loss: 0.3098 - val_accuracy: 0.9592
Epoch 12/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3041 - accuracy: 0.9607 - val_loss: 0.3117 - val_accuracy: 0.9578
Epoch 13/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3027 - accuracy: 0.9613 - val_loss: 0.3084 - val_accuracy: 0.9594
Epoch 14/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3027 - accuracy: 0.9612 - val_loss: 0.3102 - val_accuracy: 0.9585
Epoch 15/20
60936/60936 [==============================] - 79s 1ms/step - loss: 0.3019 - accuracy: 0.9616 - val_loss: 0.3076 - val_accuracy: 0.9601
Epoch 16/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3024 - accuracy: 0.9616 - val_loss: 0.3082 - val_accuracy: 0.9594
Epoch 17/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3029 - accuracy: 0.9616 - val_loss: 0.3069 - val_accuracy: 0.9600
Epoch 18/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3015 - accuracy: 0.9623 - val_loss: 0.3105 - val_accuracy: 0.9588
Epoch 19/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3012 - accuracy: 0.9624 - val_loss: 0.3138 - val_accuracy: 0.9574
Epoch 20/20
60936/60936 [==============================] - 78s 1ms/step - loss: 0.3008 - accuracy: 0.9625 - val_loss: 0.3102 - val_accuracy: 0.9588
* Accuracy on training set: 96.54%
* Accuracy on test set 1: 95.88%
* Accuracy on test set 2: 65.37%
* Accuracy on test set 3: 81.19%
</p>

